process/executionTrack/executionResultsTracker.go

```diff
@@ -89,61 +117,107 @@ func (est *executionResultsTracker) GetExecutionResultByHash(hash []byte) (*bloc
 }
 
 // GetExecutionResultsByNonce will return the execution results by nonce
-func (est *executionResultsTracker) GetExecutionResultsByNonce(nonce uint64) ([]*block.ExecutionResult, error) {
-	executionResults := make([]*block.ExecutionResult, 0)
-
+func (est *executionResultsTracker) GetExecutionResultsByNonce(nonce uint64) (*block.ExecutionResult, error) {
 	est.mutex.RLock()
 	defer est.mutex.RUnlock()
 
-	hashes := est.nonceHashes.getNonceHashes(nonce)
-	for _, hash := range hashes {
-		result, found := est.executionResultsByHash[hash]
-		if !found {
-			return nil, fmt.Errorf("%w with hash: '%s'", ErrCannotFindExecutionResult, hash)
-		}
+	return est.getExecutionResultsByNonce(nonce)
+}
 
-		executionResults = append(executionResults, result)
+func (est *executionResultsTracker) getExecutionResultsByNonce(nonce uint64) (*block.ExecutionResult, error) {
+	hash := est.nonceHash.getHashByNonce(nonce)
+	result, found := est.executionResultsByHash[hash]
+	if !found {
+		return nil, fmt.Errorf("%w with hash: '%s'", ErrCannotFindExecutionResult, hex.EncodeToString([]byte(hash)))
 	}
 
-	return executionResults, nil
+	return result, nil
 }
 
 // CleanConfirmedExecutionResults will clean the confirmed execution results
-// TODO add unit tests when new header structure is available
-func (est *executionResultsTracker) CleanConfirmedExecutionResults(headerHash []byte, header data.HeaderHandler) error {
+func (est *executionResultsTracker) CleanConfirmedExecutionResults(header HeaderWithExecutionResults) (*CleanInfo, error) {
 	est.mutex.Lock()
 	defer est.mutex.Unlock()
 
-	headerExecutionResults := make([]*block.ExecutionResult, 0) // TODO get from header
+	headerExecutionResults := header.GetExecutionResults()
 	if len(headerExecutionResults) == 0 {
-		return nil
+		return &CleanInfo{
+			CleanResult:             CleanResultOK,
+			LastMatchingResultNonce: est.lastNotarizedResult.Nonce,
+		}, nil
+	}
+
+	return est.cleanConfirmedExecutionResults(headerExecutionResults)
+}
+
+func (est *executionResultsTracker) cleanConfirmedExecutionResults(headerExecutionResults []*block.ExecutionResult) (*CleanInfo, error) {
+	pendingExecutionResult, err := est.getPendingExecutionResults()
+	if err != nil {
+		return nil, err
 	}
 
-	for _, executionResultFromHeader := range headerExecutionResults {
-		executionResultFromTracker, found := est.executionResultsByHash[string(executionResultFromHeader.HeaderHash)]
-		if !found {
-			return fmt.Errorf("%w with for header hash %s", ErrCannotFindExecutionResult, hex.EncodeToString(executionResultFromHeader.HeaderHash))
+	lastMatchingResultNonce := est.lastNotarizedResult.Nonce
+	lastMatchingHash := est.lastNotarizedResult.HeaderHash
+	for idx, executionResultFromHeader := range headerExecutionResults {
+		if idx > len(pendingExecutionResult)-1 {
+			// missing  execution result
+			return &CleanInfo{
+				CleanResult:             CleanResultNotFound,
+				LastMatchingResultNonce: lastMatchingResultNonce,
+			}, nil
```

Comment:
maybe I don't understand here completely.. shouldn't be the same extracting this check before the for loop, something like len(headerExecutionResults) > len(pendingExecutionResult)?